### PR TITLE
Fixed memory leak caused by exception.

### PR DESCRIPTION
**Pull request by limingxinleo**
fix https://github.com/guzzle/guzzle/issues/2606
fix https://github.com/hyperf/hyperf/issues/1486

### CODE REVIEW

File: src/Promise.php [hunk 1]
```diff
@@ -67,8 +67,11 @@ public function wait($unwrap = true)
             if ($this->state === self::FULFILLED) {
                 return $this->result;
             }
+            $exception = $this->result;
+            // Unset exception form Promise, because it will case memory leak.
```
Comment (mfn) [inline src/Promise.php:71]:
```suggestion
            // Unset exception from Promise otherwise it will cause a memory leak.
```

**Comment by mfn** [comment]
Interesting!

Do we know this does not happen for the `fullfilled` state too? Even in this case, there still a reference held … no?

File: src/Promise.php (hunk 1 above)
Comment (limingxinleo) [reply src/Promise.php:71]: I will change it tomorrow
Thanks

**Comment by limingxinleo** [comment]
> Interesting!
> 
> 
> 
> Do we know this does not happen for the `fullfilled` state too? Even in this case, there still a reference held … no?

If the result is not a exception, it won't cause memory leak, so fullfilled state work well.

**Comment by Reasno** [comment]
The troublesome reference was in $exception->trace (which snapshots the recursive call stack). So unless an exception is thrown, no memory leak would happen.

**Comment by Reasno** [comment]
Made a simplified demo to show what happens underhood:

```php
<?php

class MemoryLeaker
{
    private $ref;
    public function leak($self)
    {
        $this->ref = new Exception(); // Exception captures $self
    }
}


while (true) {
    echo 'memory:', memory_get_usage() / 1024 / 1024, "\r\n";
    $leaker = new MemoryLeaker();
    $leaker->leak($leaker);
    sleep(1);
}
```

**Comment by Tobion** [comment]
Hm this means you cannot call `wait` more than once anymore on the promise and still get the same result.

```
try {
    $promise->wait():
} catch (\Throwable) {
    $promise->wait(): // won't throw an exception anymore.
}
```

Isn't there another solution?

**Comment by limingxinleo** [comment]
@Tobion You are right...🤣

**Comment by GrahamCampbell** [comment]
Friendly ping @limingxinleo.

**Comment by Nyholm** [comment]
Im 👎 for the current solution because it will make repeated calls to `wait()` to have issues. 

See: 

```php
<?php

require 'vendor/autoload.php';
use GuzzleHttp\Promise\Promise;


$promise = new Promise(function () use (&$promise) {
    $promise->reject(new \RuntimeException('foobar'));
});

// Calling wait will return the value of the promise.
try {
    $promise->wait();
} catch (\Exception $e) {
    echo $e->getMessage();
}

try {
    $promise->wait();
} catch (\Exception $e) {
    echo $e->getMessage();
}
```

This will print: 

```txt
# On master
foobar
foobar

# On this PR
foobar
The promise was rejected

```

**Comment by GrahamCampbell** [comment]
Yeh, I don't think we can merge this, as it. We will need to simulate weak references, I think, to actually fix this?

**Comment by Nyholm** [comment]
Here is a fix that will work: https://github.com/guzzle/promises/pull/128

**Pull request closed without merging**